Proofread the README (#582)
tarantool/cartridge-cli: README.rst

.. _cartridge-cli:

Cartridge Command Line Interface
================================

.. image:: https://img.shields.io/github/v/release/tarantool/cartridge-cli?include_prereleases&label=Release&labelColor=2d3532
   :alt: Cartridge CLI latest release on GitHub

>    :target: https://github.com/tarantool/cartridge-cli/releases

Reviewer: The same badge is just below (but with 'Cartridge-CLI' instead of 'Cartridge CLI').

Now enable shell completion:

.. code-block:: bash

    echo "autoload -U compinit; compinit" >> ~/.zshrc

OS X
~~~~

If you install ``cartridge-cli`` from ``brew``, it automatically installs both
Bash and Zsh completion.

Usage
-----

For more details, use the ``--help`` flag:

.. code-block:: bash

   cartridge --help

Here is a list of supported Cartridge CLI commands:

* ``create``: create a new application from template.
* ``build``: build an application for local development and testing.
* ``start``: start one or more Tarantool instances.
* ``stop``: stop one or more Tarantool instances.
* ``status``: get the status of one or more current instances.
* ``log``: get logs for one or more instances.
* ``clean``: clean files for one or more instances.
* ``pack``: pack the application into a distributable bundle.
* ``repair``: patch cluster configuration files.
* `admin <doc/admin.rst>`__: call an admin function provided by the application.
* `replicasets <doc/replicasets.rst>`__: manage cluster replica sets running locally.
* `enter <doc/connect.rst>`__ and `connect <doc/connect.rst>`__: connect to a running instance.

> * `failover <doc/failover.rst>`_: manage cluster failover.

Reviewer: 1. Why one underscore is changed to two underscores (I don't know RST well enough, sorry)? I see no difference on the document rendered by GitHub.
2. The last item still has one underscore: typo?

You can control output verbosity with these global flags:

* ``verbose``: displays additional log messages as well as
  commands/docker output, such as the output of ``tarantoolctl rocks make`` or ``docker build``.
* ``debug``: works the same as verbose, but temporary files and
  directories aren't removed during command execution.
* ``quiet``: hides all logs, only displays error messages.

Application lifecycle
~~~~~~~~~~~~~~~~~~~~~

In a nutshell:

1. :ref:`Create an application <cartridge-cli-creating_an_application_from_template>`

>    (for example, ``myapp``) from a template:

Reviewer: The link looks weird (and does not work) when rendered by GitHub. It is highly undesirable in a README file. The same for all `:ref:` and `:doc:` usages. Maybe also for other templates specific for our website. Please, re-check how the whole document looks on GitHub.

![image](https://user-images.githubusercontent.com/92749/132384538-bcca1a1a-26f4-4393-9606-7bfb69e1485e.png)

Reviewer: I guess you meant 'the template' generated by the tool, not some arbitrary template.

Reviewer: > I guess you meant 'the template' generated by the tool, not some arbitrary template.

At least, we don't say anything about `--dir` at this point of the document.

   .. code-block:: bash

       cartridge create --name myapp
       cd ./myapp

2. :ref:`Build the application <cartridge-cli-building-the-application>`
   for local development and testing:

   .. code-block:: bash

       cartridge build

3. :ref:`Run instances locally <cartridge-cli-starting-the-application-locally>`:

   .. code-block:: bash

       cartridge start
       cartridge stop

4. :ref:`Pack the application <cartridge-cli-packaging-the-application>`
   into a distributable (like an RPM package):

   .. code-block:: bash

       cartridge pack rpm

.. _cartridge-cli-creating_an_application_from_template:

Creating an application from a template
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To create an application from a Cartridge template, run ``cartridge create`` in any directory:

Reviewer: From arbitrary Cartirdge template? I guess the sentence is about certain template, generated by cartridge cli.

Reviewer: Here I'm in doubt. We didn't mention that there is a way to choose a custom template, but it is in several lines below. So up to you.

..  code-block:: bash

    cartridge create [PATH] [flags]

Supported options (``[flags]``):

.. // Please update the doc in cli/commands on updating this section

*   ``--name string``: application name.

*   ``--from DIR``: path to the application template. For more about templates, see below.

*   ``--template string``: name of the application template to be used.
    Only the template ``cartridge`` is supported at the time.

> Your application will appear in the ``<path>/<app-name>/`` directory.

Reviewer: It is named just `<app_name>/` below. I'm for consistency (I would prefer just `<app_name>/`).

* ``log`` – get logs for one or more instances.
* ``clean`` – clean files for one or more instances.
* ``pack`` – pack the application into a distributable bundle.
* ``repair`` – patch cluster configuration files.
* `admin <doc/admin.rst>`__ – call an admin function provided by the application.
* `replicasets <doc/replicasets.rst>`__ – manage cluster replica sets running locally.
* `enter <doc/connect.rst>`__ and `connect <doc/connect.rst>`__ – connect to a running instance.

You can control output verbosity with these global flags:

* ``verbose`` – verbose mode: displays additional log messages as well as
  commands/docker output, such as the output of `tarantoolctl rocks make` or `docker build`
* ``debug`` – debug mode: works the same as verbose, but temporary files and
  directories aren't removed during command execution.
* ``quiet`` – hides all logs, only displays error messages.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Application lifecycle
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In a nutshell:

1. :ref:`Create an application <cartridge-cli-creating_an_application_from_template>`

>     (for example, ``myapp``) from a template:

Reviewer 2: OK

   .. code-block:: bash

       cartridge create --name myapp
       cd ./myapp

2. :ref:`Build the application <cartridge-cli-building-an-application>`
   for local development and testing:

   .. code-block:: bash

       cartridge build

3. :ref:`Run instances locally <cartridge-cli-starting-stopping-an-application-locally>`:

   .. code-block:: bash

       cartridge start
       cartridge stop

4. :ref:`Pack the application <cartridge-cli-packing-an-application>`

>    into a distributable (for instance, an RPM package):

Reviewer 2: ```suggestion
   into a distributable (like an RPM package):
```

.. _cartridge-cli-creating_an_application_from_template:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Creating an application from a template
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create an application from a Cartridge template, run ``cartridge create`` in any directory:

..  code-block:: bash

    cartridge create [PATH] [flags]

Supported options (``[flags]``):

.. // Please update the doc in cli/commands on updating this section

> * ``--name string`` – application name.

Author: Dashes are hard to discern when one's looking at a diff in monospace font. Instead of explicit m- and n-dash characters we can use triple and double hyphens:

```suggestion
*    ``--name string`` --- application name.
```

Reviewer 2: Aren't we using n-dashes in English texts? If we are, then we have to use a double dash `--`, correct?

> * ``--from DIR`` – path to the application template. For more about templates, see below.

Author: ```suggestion
*    ``--from DIR`` – path to the application template. For more about templates, see below.
```

Reviewer 2: done

> * ``--template string`` – name of the application template to be used.
>   Only the template ``cartridge`` is supported at the time.

Author: indentation matters
```suggestion
*   ``--template string`` – name of the application template to be used.
     Only the template ``cartridge`` is supported at the time.
```

Your application will appear in the ``<path>/<app-name>/`` directory.

If you have ``git`` installed, a Git repository with
a ``.gitignore`` file will be also set up in the project root directory.
The initial commit will be created, tagged with
`version <https://www.tarantool.io/en/doc/latest/book/cartridge/cartridge_dev/#application-versioning>`_
0.1.0.

^^^^^^^^^^^^^^^^^
Project directory
^^^^^^^^^^^^^^^^^

> Let's take a closer look at the files inside the ``<app_name>/`` directory.

Author: Sounds good )

* Application files:

  * ``app/roles/custom-role.lua`` is a sample
    `custom role <https://www.tarantool.io/en/doc/latest/book/cartridge/cartridge_dev/#cluster-roles>`_
    with a simple HTTP API. Can be enabled as ``app.roles.custom``.
  * ``<app_name>-scm-1.rockspec`` contains application
    dependencies.
  * ``init.lua`` is the application entry point.
  * ``stateboard.init.lua`` is the application

>     `stateboard <https://github.com/tarantool/cartridge/blob/master/rst/topics/failover.rst>`__ entry point.

Author: https://www.tarantool.io/en/doc/latest/book/cartridge/cartridge_dev/#failover-architecture
```suggestion
    :ref:`stateboard <cartridge-failover>` entry point.
```

Reviewer 2: Also made this a `:doc:` link as per the [SG](https://www.tarantool.io/en/doc/latest/contributing/docs/markup/links/#linking-to-other-documentation-pages).

> * `Build and packing files <Special files_>`_:

Author: bad link, replace with anchor+ref

      │   └── unit
      │       └── sample_test.lua

* Configuration files:

  * ``.luacheckrc``
  * ``.luacov``
  * ``.editorconfig``

Using a custom template
^^^^^^^^^^^^^^^^^^^^^^^

The template used by default is ``cartridge``.
It produces a simple Cartridge application that includes:

* One custom role with an HTTP endpoint.
* Sample tests and basic test helpers.
* Files required for development (like ``.luacheckrc``).

To create an application based on your own custom template, run ``cartridge create`` with the ``--from`` flag.

If the template directory is a Git repository, all files in the ``.git`` directory will be ignored on
instantiating the template.
Instead, a new git repo will be initialized for the newly created application.

Don't include the ``.rocks`` directory in your template application.
To specify application dependencies, use the ``.rockspec`` and ``cartridge.pre-build`` files.

Text variables

> ^^^^^^^^^^^^^^

Reviewer: Should not it be a subsection of 'Using a custom template'?

      │   └── unit
      │       └── sample_test.lua

* Configuration files:

  * ``.luacheckrc``
  * ``.luacov``
  * ``.editorconfig``

^^^^^^^^^^^^^^^^^^^^^^^
Using a custom template
^^^^^^^^^^^^^^^^^^^^^^^

The template used by default is ``cartridge``.
It produces a simple Cartridge application that includes:

* One custom role with an HTTP endpoint.
* Sample tests and basic test helpers.
* Files required for development (like ``.luacheckrc``).

To create an application based on your own custom template, run ``cartridge create`` with the ``--from`` flag.

> If the template directory is a Git repository, all files in the `.git` directory will be ignored on

Author: ```suggestion
If the template directory is a Git repository, all files in the ``.git`` directory will be ignored on
```

instantiating the template.
Instead, a new git repo will be initialized for the newly created application.

> Don't include a `.rocks` directory in your template application.

Author: ```suggestion
Don't include the ``.rocks`` directory in your template application.
```

> To specify application dependencies, use rockspec and `cartridge.pre-build` files.

Author: ```suggestion
To specify application dependencies, use ``.rockspec`` and ``cartridge.pre-build`` files.
```

.. code-block:: lua

    print("Hi, I am the {{ .Name }} application")
    print("I also have a stateboard named {{ .StateboardName }}")

.. _cartridge-cli-building-the-application:

Building the application
~~~~~~~~~~~~~~~~~~~~~~~~

To build your application locally (for local testing), run this in any directory:

.. code-block:: bash

    cartridge build [PATH] [flags]

The following options (``[flags]``) are supported:

* ``--spec`` is the path to the ``.rockspec`` to use for the current build.
  *Note* that the ``.rockspec`` file name should be in lowercase.

.. // Please update the doc in cli/commands on updating this section

The command requires one argument---the path to your application directory
(that is, to the build source).
The default path is ``.`` (the current directory).

``cartridge build`` is executed in two steps:

> 1. If there is a  :ref:`pre-build file <cartridge-cli-special-files>`, ``cartridge.pre-build`` builds the application in the ``[PATH]`` directory.

Reviewer: Can we name it somehow different? Say, `DESTDIR`? I read it several times until got that it is the `PATH` from the command declaration above, not how it is usually used (the environment variable with executables paths).

Reviewer: 'cartridge.pre-build build the application' — it looks as a mistake.

Reviewer: So, we call the prebuild script and then build the app in the DESTDIR. Correct? If so, please, change the description. Now I cannot undesrtand it.

2. If there is a :ref:`rockspec file <cartridge-cli-special-files>`, ``tarantoolctl rocks make`` installs all Lua rocks to the ``[PATH]`` directory.

> First, ``cartridge`` builds the application inside the application directory.

Reviewer: Does not it contradict with the statement above: 'This builds the application in the ``[PATH]`` directory'?

rocks from a project added as a submodule, install all those
dependencies **before** calling ``tarantoolctl rocks make``.
You can do so using the file ``cartridge.pre-build`` in your application root.
That file is also located in the application directory created from template.
In ``cartridge.pre-build``, you can specify all the rocks to build
(for example, add ``tarantoolctl rocks make --chdir ./third_party/proj``).
For details, see :ref:`build and packaging files <cartridge-cli-special-files>`.

As a result, a fully built application will appear in the ``.rocks`` directory.
You can start it locally from the application directory.

.. _cartridge-cli-starting-the-application-locally:

Starting the application locally
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After you've :ref:`built your application <cartridge-cli-building-the-application>`,
you can run it locally:

.. code-block:: bash

    cartridge start [INSTANCE_NAME...] [flags]

where ``[INSTANCE_NAME...]`` stands for one or multiple instance names.

If no ``INSTANCE_NAME`` is provided, all the instances in the
Cartridge instances configuration file will be taken as arguments (see the ``--cfg``
option below).

> .. // TODO add links on every "below"

Reviewer: I don't get it.

The application name, ``APP_NAME``, is passed to the instances during startup and
used in instance file paths,

> for example: ``<run-dir>/<APP_NAME>.<INSTANCE_NAME>.pid``).

Reviewer: Whether the `<run-dir>` is the current directory or some another specially created directory? AFAIS, it is not mentioned anywhere above.

Reviewer: Ah, I see. It is below. Not nice, but at least does not incorrect.

call.

Example ``instances.yml`` file:

..  code-block:: yaml

    myapp.router:
        advertise_uri: localhost:3301
        http_port: 8081

    myapp.s1-master:
        advertise_uri: localhost:3302
        http_port: 8082

For the full list of parameters that can be specified in ``instances.yml``, read the
:doc:`cartridge.cfg() documentation </book/cartridge/cartridge_api/modules/cartridge/#cfg-opts-box-opts>`.
``advertise_uri`` is a required parameter.

..  note::

    The following parameters, if specified in ``instances.yml``, will be
    overwritten by Cartridge CLI environment variables on
    ``cartridge start``:

    * ``workdir``
    * ``console_sock``
    * ``pid_file``.

You can specify custom parameters in ``instances.yml``, but they also
have to be defined in your application code. See the following example, where
``my_param`` is a custom parameter:

``instances.yml``:

..  code-block:: yaml

    myapp.router:
        advertise_uri: localhost:3301
        http_port: 8081
        my_param: 'Hello, world'

``init.lua``:

..  code-block:: lua

    local argparse = require('cartridge.argparse')

>     local my_param = argparse.get_opts({my_param='string'} )

Reviewer: Please, remove the whitespace before the closing parenthesis.

> * ``--daemonize, -d`` starts the instance(s) in the background.

Reviewer: You formatted some options into the table above, but left some (like this one) in a list. Now it is really hard to understand that all of them are the same list and could be passed to `cartridge start`.

  With this option, Tarantool also waits until the application's init script
  finishes evaluating.
  This is useful if ``init.lua`` requires time-consuming startup
  from a snapshot. Another use case would be if your application's init script

>   generates errors, so Tarantool can handle them.

Reviewer: 'Tarantool can handle them' is unlikely a result of 'init script generates errors'. To be honest, I don't understand how an uncatched Lua error from the main script could be hanlded except exitting with a non-zero exit code.

Building an application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To build your application locally (for local testing), run this in any directory:

.. code-block:: bash

    cartridge build [PATH] [flags]

.. // Please update the doc in cli/commands on updating this section

> The command requires one argument – the path to your application directory
> (that is, to the build source). The default path is ``.`` (the current directory).

Author: Whenever you can, start new sentences in the source files from a new line. It makes sources better readable. 
See https://www.tarantool.io/en/doc/latest/contributing/docs/markup/intro/#wrapping-text.
```suggestion
The command requires one argument – the path to your application directory
(that is, to the build source).
The default path is ``.`` (the current directory).
```

`cartridge build` is executed in two steps:

1. If there is a  `pre-build file <Special files_>`_, ``cartridge.pre-build`` builds the application in the ``[PATH]`` directory.
2. If there is a `rockspec file <Special files_>`_, ``tarantoolctl rocks make`` installs all Lua rocks to the ``[PATH]`` directory.

First, ``cartridge`` builds the application inside the application directory.
This is different from ``cartridge pack``, which builds the application inside the
`build directory <Build directory_>`_. No build artifacts
remain in the application directory.

> Second, ``cartridge`` installs all dependencies specified in the rockspec file.

Author: If it's a certain file created from template, we probably know the exact name. At least, it should be `something.rockspec`.

```suggestion
Second, ``cartridge`` installs all dependencies specified in the `.rockspec` file.
```

(That file is located in the application directory created from template.)

Alternatively, you can define the build logic in the rockspec in the form of ``cmake`` commands,
`like we do <https://github.com/tarantool/cartridge/blob/master/cartridge-scm-1.rockspec#L26>`_

> in ``cartridge``.)

Author: We do it in the Cartridge software project in general, not in the `cartridge` as a CLI application.

```suggestion
in Cartridge.)
```

If your application depends on closed-source rocks, or if your build contains
rocks from a project added as a submodule, install all those
dependencies **before** calling ``tarantoolctl rocks make``.

> You can do so using the file ``cartridge.pre-build`` in your application root
> (again, that file is located in the application directory created from template).

Author: ```suggestion
You can do so using the file ``cartridge.pre-build`` in your application root.
That file is also located in the application directory created from a template.
```

Reviewer 2: I argue the indefinite article is out of place here, as we have been speaking about that application directory and the template it's based upon for a long while already. "From template" is pretty googlable (3 million results). Alternatively, we can say "from your template".

In this file, you can specify all the rocks to build
(e.g. ``tarantoolctl rocks make --chdir ./third_party/proj``).

> For details, see `build and packing files <Special files_>`_.

Author: bad link, replace with anchor+ref

As a result, a fully built application will appear in the ``.rocks`` directory.

> You can start it locally from the application directory.

Author: Good )

.. _cartridge-cli-starting-stopping-an-application-locally:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting/stopping an application locally
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> *********

Author: We don't use `*` for headings. https://www.tarantool.io/en/doc/latest/contributing/docs/markup/headings/

``start``
*********

> After you've `built <Building an application_>`_ your application,

Author: bad link :)

you can run it locally:

.. code-block:: bash

    cartridge start [INSTANCE_NAME...] [flags]

where ``[INSTANCE_NAME...]`` stands for one or multiple instance names.

If no ``INSTANCE_NAME`` is provided, all the instances in the
Cartridge instances configuration file will be taken as arguments (see the ``--cfg``
option below).

.. // TODO add links on every "below"

> The application name, ``APP_NAME``, is passed to the instances during startup and
> used in instance file paths (for example, ``<run-dir>/<APP_NAME>.<INSTANCE_NAME>.pid``).

Author: ```suggestion
The application name, ``APP_NAME``, is passed to the instances during startup and
used in instance file paths.
For example, ``<run-dir>/<APP_NAME>.<INSTANCE_NAME>.pid``.
```

Reviewer 2: The last sentence looks truncated, I would reformulate:
```suggestion
The application name, ``APP_NAME``, is passed to the instances during startup and
used in instance file paths, like in the following example: ``<run-dir>/<APP_NAME>.<INSTANCE_NAME>.pid``.
```

> By default, ``APP_NAME`` is derived from the application rockspec in the current
> directory. However, the variable also can be defined explicitly via the ``--name`` option
> (see below).

Author: It's better to wrap lines by sentences.

```suggestion
By default, ``APP_NAME`` is derived from the application rockspec in the current directory.
However, the variable also can be defined explicitly via the ``--name`` option (see below).
```

^^^^^^^^
Options
^^^^^^^^

Supported options (``[flags]``):

.. // Please update the doc in cli/commands on updating this section

* ``--script FILE`` is the application's entry point.
  Can be either an absolute or a relative path to the entry point in the project directory.
  Defaults to ``init.lua`` or the value of the ``script``

>   parameter in the Cartridge `configuration file <Overriding default options_>`_.

Author: a few wrong links here )

        advertise_uri: localhost:3302
        http_port: 8082

  For a full list of parameters that can be specified in ``instances.yml``, read the

>   `cartridge.cfg() documentation <https://www.tarantool.io/en/doc/latest/book/cartridge/cartridge_api/modules/cartridge/#cfg-opts-box-opts>`_.

Author: this should be a `:ref:` link

Reviewer 2: I made this a `:doc:` link according to [the SG](https://www.tarantool.io/en/doc/latest/contributing/docs/markup/links/#linking-to-other-documentation-pages).

    script: my-init.lua
    stateboard: true

**Note:** the config of the `standard application template <Creating an application from template_>`_
initially has the ``stateboard`` parameter set to ``true``.

.. // Please update the doc in cli/commands on updating this section

> ********

Author: Headings can't have formatted content, like inline code. Only plain text. Also, `*` is a wrong character, just like above.

``stop``
********

> To stop one or more running instances, execute:

Author: we usually use 'run'.

Reviewer 2: "...running instances, run" looks less than ideal

.. code-block:: bash

    cartridge stop [INSTANCE_NAME...] [flags]

By default, the instances receive a SIGTERM.

Supported options (``[flags]``):

> * ``-f, --force`` allows force-stopping the instance(s) by a SIGKILL.

Author: I think it's 'with', because SIGKILL is just a number, not some program or tool.

Reviewer 2: OK

.. note::

>    When you call ``cartridge stop``, use the exact same ``run-dir`` as When

Author: ```suggestion
   When you call ``cartridge stop``, use the exact same ``run-dir`` as when
```

.. code-block:: bash

    cartridge status [INSTANCE_NAME...] [flags]

`Some options <Options_>`_ are identical to those of the ``start`` command:

* ``--run-dir DIR``
* ``--cfg FILE``
* ``--stateboard``
* ``--stateboard-only``

.. // Please update the doc in cli/commands on updating this section

*******
``log``
*******

Use the ``log`` command to display the logs of one or more instances running in the background:

.. code-block:: bash

    cartridge log [INSTANCE_NAME...] [flags]

Supported options (``[flags]``):

* ``-f, --follow`` outputs appended data as the log grows.

* ``-n, --lines int`` is the number of last lines to be displayed.
  Defaults to 15.

`Some options <Options_>`_ are identical to those of the ``start`` command:

* ``--log-dir DIR``
* ``--run-dir DIR``
* ``--cfg FILE``
* ``--stateboard``
* ``--stateboard-only``

.. // Please update the doc in cli/commands on updating this section

*********
``clean``
*********

Use the ``clean`` command to remove the files associated with one or more instances
(such as the log file, workdir, console socket, PID file and notify socket):

.. code-block:: bash

    cartridge clean [INSTANCE_NAME...] [flags]

> Executing `cartridge clean` for running instance(s) causes an error.

Author: ```suggestion
Executing ``cartridge clean`` for running instance(s) causes an error.
```

> `Some options <Options_>`_ are identical to those of the ``start`` command:

Author: Baaad way to use a link. In translation, "Options" will become "Опции" and the link won't work. Use explicit anchors instead:

```
..  _cartridge-cli-options:

Options
--------
```

And change the link to:

```suggestion
:ref:`Some options <cartridge-cli-options>` are identical to those of the ``start`` command:
```

Reviewer 2: done

* ``PATH`` (optional) is the path to the application directory that you want to pack.
  Defaults to ``.`` (the current directory).

.. note::

  If you pack your application into an RPM or DEB on MacOS without the ``--use-docker``
  flag, the final artifact will be broken, because it will contain rocks and executables
  that can't be used on Linux. In this case packing will fail.

Supported options (``[flags]``):

.. // Please update cmd_pack usage in cartridge-cli.lua file on updating the doc

* ``--name string`` (common for all distribution types) is the application name.
  It is the same as the package name and the systemd service name.
  By default, it is derived from the ``package`` field in the ``.rockspec`` file.

* ``--version string`` (common for all distribution types) is the package version.
  The expected pattern is ``major.minor.patch[-count][-commit]``.
  If you specify the version in the ``major.minor.patch``format,
  it will be normalized to ``major.minor.patch-count``.
  By default, the version string is the result of ``git describe --tags --long``.
  If your application is not a git repository, you have to set the ``--version`` option
  explicitly.

> * ``--suffix string`` (common for all distribution types) is the suffix of the

Reviewer 2: Is suffix the same as extension?

Reviewer 3: нет. например пакет rpm у нас именуется так: ``[name]-[version]-[release].[arch].rpm``. 
С опцией суффикса становится так: ``[name]-[version]-[release]-[suffix].[arch].rpm``

* ``--instantiated-unit-template string`` (for ``rpm`` and ``deb`` packages only)
  is the path to the template for the ``systemd`` instantiated unit file.

* ``--stateboard-unit-template string`` (for ``rpm`` and ``deb`` packages only)
  is the path to the template for the stateboard ``systemd`` unit file.

* ``--use-docker`` (enforced for ``docker``) forces to build the application in Docker.

* ``--tag strings`` (for ``docker`` only) is the tag(s)
  of the Docker image that results from ``pack docker``.

* ``--from string`` (for ``docker`` only) is the path to the base Dockerfile of the runtime
  image. Defaults to ``Dockerfile.cartridge`` in the application directory.

* ``--build-from string`` (common for all distribution types, used for building in Docker)
  is the path to the base Dockerfile of the build image.
  Defaults to ``Dockerfile.build.cartridge`` in the application directory.

* ``--no-cache`` creates build and runtime images with the ``--no-cache`` Docker flag.

* ``--cache-from strings`` are the images that work as cache sources for both build and
  runtime images. See the ``--cache-from`` flag for the ``docker build`` command.

> 

Reviewer 2: I believe it's a good idea to put looong lists like this in the form of a table. This will make it easier to read. What do you think, @NickVolynkin?

* ``--sdk-path string`` (common for all distribution types, used for building in Docker)

>   is the path to the SDK that will be delivered in the final artifact.

Reviewer 2: Does "delivered" mean "included in the final package"?

Reviewer 3: да

    dependency_03==2
    dependency_04<5,>=1.5.3

Each line must describe a single dependency.
You can specify both the major and minor version of the dependency:

.. code-block:: bash

    dependency_05 >= 4, < 5

The ``--deps`` and ``--deps-file`` flags require similar formats of dependency information.
However, ``--deps`` does not allow you to specify major and minor versions:

.. code-block:: bash

    # You can't do that:
    cartridge pack rpm --deps dependency_06>=4,<5 appname

    # Instead, do this:
    cartridge pack rpm --deps dependency_06>=4,dependency_06<5 appname

    # Or this:
    cartridge pack rpm --deps dependency_06>=4 --deps dependency_06<5 appname

For Tarantool Enterprise, specify either ``--sdk-local`` or ``--sdk-path``
(not both at the same time).

> For ``rpm``, ``deb``, and ``tgz``, we also deliver rocks modules and executables

Reviewer 2: What exactly does "we also deliver" mean in this context? What does the user have to do to get their hands on those rocks and executables?
Also, is it correct to say "rocks modules"? I'm also considering options like "rocks", "rock modules" and "`rocks` modules".

Reviewer 3: we also deliver - вставляем в пакет папку или бинарник

Reviewer 3: я думаю lua rocks или просто rocks

Reviewer 2: thank you!

delete the ``~/.cartridge/tmp/cache`` directory.

Next, let's dive deeper into the packaging process.

.. _cartridge-cli-build-directory:

Build directory
^^^^^^^^^^^^^^^

The first step of the packaging process is to
:ref:`build the application <cartridge-cli-building-the-application>`_.

By default, the application is built inside a temporary directory in
``~/.cartridge/tmp/``, so the packaging process doesn't affect the contents
of your application directory.

You can specify a custom build directory for your application in the ``CARTRIDGE_TEMPDIR``
environment variable. If this directory doesn't exist yet, it will be created, used
for building the application, and then removed.

If you specify an existing directory in the ``CARTRIDGE_TEMPDIR`` environment
variable, the ``CARTRIDGE_TEMPDIR/cartridge.tmp`` directory will be used for
building and then removed. This directory will be cleaned up before building the
application.

.. _cartridge-cli-distribution-directory:

Distribution directory
^^^^^^^^^^^^^^^^^^^^^^

For each distribution type, a temporary directory with application source files
is created (further on we address it as *application directory*).

> This includes 3 stages.

Reviewer 2: мне кажется, абзац вносит небольшую путаницу, и я бы хотела дальше говорить distribution directory, а не application directory

Distribution directory - это ведь не то же самое, что корневая папка приложения (то, что выше называлось application directory)? это разные вещи? just to be clear

на что указывает This в This includes 3 stages - создание этой директории или упаковка пакета?
до этого у нас упоминается, что The first step of the packaging process is to build the application
то есть перед стадиями 1, 2 и 3 возникают ещё стадии...

Reviewer 3: в этом абзаце сейчас всё достаточно запутано и я думаю его нужно переделать

> Build directory

тут речь о том, что билд (на самом деле 3 стадии из **Distribution  directory**) происходят во временной директории, и ниже описываются правила, по которым она задается (*By default, application build is done...*)

> Distribution directory

создается временная папка по правилам выше и в ней выполняются три стадии описанные ниже (+ перед этим туда копируется приложение). 
а потом уже выполняется непосредственно упаковка - вызов rpmbuild для rpm, docker для docker image и так далее.

Author: @mRrvz спасибо за комменты, мы их учтём, но не в этом ПР. Тут скоуп задачи — только переписать текст, чтобы улучшить язык и оформление, но не менять смысл и содержание. (Иначе задача будет бесконечно большой)

Reviewer 2: > в этом абзаце сейчас всё достаточно запутано и я думаю его нужно переделать
> 
> > Build directory
> 
> тут речь о том, что билд (на самом деле 3 стадии из **Distribution directory**) происходят во временной директории, и ниже описываются правила, по которым она задается (_By default, application build is done..._)
> 
> > Distribution directory
> 
> создается временная папка по правилам выше и в ней выполняются три стадии описанные ниже (+ перед этим туда копируется приложение).
> а потом уже выполняется непосредственно упаковка - вызов rpmbuild для rpm, docker для docker image и так далее.

значит ли это, что Build directory и Distribution directory - одно и то же, та самая временная директория?

(мы всё-таки решили довести текст до ума, насколько это возможно сделать быстро)

Reviewer 2: сюда же вопрос к [этим строчкам](https://github.com/tarantool/cartridge-cli/pull/582/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR957):

If you specify an existing directory in the ``CARTRIDGE_TEMPDIR`` environment
variable, the ``CARTRIDGE_TEMPDIR/cartridge.tmp`` directory will be used for
building and then removed. This directory will be cleaned up before building the
application.

какая часть билда выполняется в cartridge.tmp, если эта директория удаляется ещё перед билдом? мне кажется, здесь есть какая-то неточность

files.
In default mode, ``repair`` walks through all cluster-wide configurations
in ``<data-dir>/<app-name>.*`` directories and patches all the configuration
files it locates.

If the ``--dry-run`` flag is specified, files aren't patched, and only a computed
configuration diff is shown.
With the ``--dry-run`` flag specified, files won't be patched,
and you will only see the computed configuration diff.

If configuration files are diverged between instances on the local machine,
If configuration files differ between instances on the local machine,
``repair`` raises an error.
But you can specify the ``--force`` option to patch different versions of
configuration independently.
To patch different versions of configuration independently,
specify the ``--force`` option.

``repair`` can also reload configuration for all instances if the ``--reload``
flag is specified (only if the application uses ``cartridge >= 2.0.0``).
If your application uses ``cartridge >= 2.0.0``,
you can also run ``repair`` with the ``--reload`` flag
to reload configuration for all your instances.
Configuration will be reloaded for all instances that are placed in the new

> configuration using console sockets that are placed in the run directory.

Reviewer 2: Kindly clarify the meaning of this string. The sentence sounds vague now.

Reviewer 3: если указать опцию ``--reload``, конфигурация всех инстансов обновится (перезагрузится), используя консольные сокеты, которые находятся в tmp/run (*run directory*)

Reviewer 2: спасибо!
разве run directory - не `/var/run/tarantool/<app-name>`? так написано [ниже](https://github.com/tarantool/cartridge-cli/pull/582/files/0c4e4bde532a8f2b7c7524fce3d4f58a7f4f0fbd#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fL1536)

Reviewer 3: да, я перепутал

The installed package name will be ``<name>`` no matter what the artifact name is.

It contains meta information: the package name (which is the application name),
and the package version.
The package contains metadata, specifically its name (which is the application name)
and version.

If you use an opensource version of Tarantool, the package has a ``tarantool``
If you use an open source version of Tarantool, the package has a ``tarantool``
dependency (version >= ``<major>.<minor>`` and < ``<major+1>``, where
``<major>.<minor>`` is the version of Tarantool used for packing the application).
You should enable the Tarantool repo to allow your package manager install
this dependency correctly:
``<major>.<minor>`` is the version of Tarantool used for packaging the application).
Enable the Tarantool repo so that your package manager installs the dependency correctly:

* for both RPM and DEB:

  .. code-block:: bash

      curl -L https://tarantool.io/installer.sh | VER=${TARANTOOL_VERSION} bash

The package contents is as follows:
After unpacking, the contents of the package are placed in specific locations:

* the contents of the distribution directory, placed in the

>   ``/usr/share/tarantool/<app-name>`` directory

Reviewer 2: [выше](https://github.com/tarantool/cartridge-cli/pull/582/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fL979) сказано, что distribution directory временная. почему она тогда осталась в пакете?

Reviewer 3: distribution directory сверху упаковывается в пакет, а на локальной машине (где была упаковка)  distribution directory  удаляется

                    -p 127.0.0.1:8082:8082 \
                    myapp:1.0.0

To check the instance logs, run:

.. code-block:: bash

    docker logs instance-1

Runtime image tag
^^^^^^^^^^^^^^^^^

The final image is tagged as follows:

* ``<name>:<detected_version>[-<suffix>]`` by default.
* ``<name>:<version>[-<suffix>]`` if the ``--version`` parameter is specified.
* ``<tag>`` if the ``--tag`` parameter is specified.

.. _cartridge-cli-build-and-runtime-images:

> Build and runtime images

Reviewer 2: The reader will only know what build and runtime images are when reaching line 1448 of the text. The terms "build image" and "runtime image" have already been mentioned multiple times. This is confusing, so the text needs restructuring. What do you think, @NickVolynkin?

machines, because it will be used by all the other instances to connect to this
one. For example, if you start an instance with an advertise URI set to
``localhost:3302``, and then address it as ``<instance-host>:3302`` from other
instances, this won't work: the other instances will be recognizing it only as
``localhost:3302``.
Each instance's ``advertise_uri`` must be the same on all machines,
because all other instances use it to connect to that instance.
Suppose that you start an instance with ``advertise_uri`` set to
``localhost:3302``. Addressing that instance as ``<instance-host>:3302`` from a different
instance won't work, because other instances will only recognize it as ``localhost:3302``.

If you specify only a port, ``cartridge`` will use an auto-detected IP,
so you need to configure Docker networks to set up inter-instance communication.
If you specify only a port, ``cartridge`` will use an auto-detected IP.
In this case you have to configure Docker networks to set up inter-instance communication.

You can use Docker volumes to store instance snapshots and xlogs on the
host machine. To start an image with a new application code, just stop the

> old container and start a new one using the new image.

Reviewer 2: "To start an image... start a new container using the new image."
Something is illogical here. Kindly clarify the meaning of this string, I am not sure I interpreted it correctly in my recent changes.

Reviewer 3: > You can use Docker volumes to store instance snapshots and xlogs on the host machine. 
 
речь о том что можно хранить логи и снапшоты на локальной машине с помощью докер volume

> To start an image with a new application code, just stop the old container and start a new one using the new image.

захотели поменять код приложения в новом имадже? не проблема, состояния инстансов сохраняются у нас на машине (см мой комментарий выше). можно просто сделать новый имадж, остановить контейнер, запустить новый контейнер на новом имадже - а все данные самостоятельно подгрузятся из  докер volume

Reviewer 3: хорошо
